ocfweb/stats/printing.py

```diff
@@ -65,10 +66,67 @@ def _semester_histogram():
 
 @periodic(3600)
 def _toner_changes():
-    return [
-        (printer, _toner_changes_for_printer(printer))
-        for printer in ALL_PRINTERS
-    ]
+    return [(printer, _toner_used_by_printer(printer)) for printer in ACTIVE_PRINTERS]
+
+
+# Toner numbers can be significantly noisy, including significant diffs
+# whenever toner gets taken out and put back in whenever there is jam. Because
+# of this it's hard to determine if a new toner is inserted into a printer to
+# reduce this noise we only count diffs that are smaller than a cutoff which
+# empirically seems to be more accurate
+def _toner_used_by_printer(printer, cutoff=.05, since=date(2017, 8, 20)):
+    with stats.get_connection() as cursor:
+        cursor.execute(
+            '''
+            CREATE TEMPORARY TABLE ordered1
+                (PRIMARY KEY (position))
+                AS (
+                    SELECT * FROM (
+                        SELECT
+                            T.*,
+                            @rownum := @rownum + 1 AS position
+                            FROM (
+                                (
+                                    SELECT * FROM printer_toner_public
+                                    WHERE printer = %s AND
+                                    date > %s
+                                    ORDER BY date
+                                ) AS T,
+                                (SELECT @rownum := 0) AS r
+                            )
+                    ) AS x
+                )
+        ''', (printer, since.strftime('%Y-%m-%d')),
+        )
+        cursor.execute('''
+            CREATE TEMPORARY TABLE ordered2
+                (PRIMARY KEY (position))
+                AS (SELECT * FROM ordered1)
+        ''')
+        cursor.execute('''
+            CREATE TEMPORARY TABLE diffs
+            AS (SELECT
+                B.date AS date,
+                A.value/A.max - B.value/B.max as pct_diff
+                FROM
+                    ordered1 as A,
+                    ordered2 as B
+                WHERE
+                    B.position = A.position + 1)
+        ''')
+        cursor.execute(
+            '''
+            SELECT SUM(pct_diff) as toner_used
+            FROM
+            diffs
+            WHERE
+            ABS(pct_diff)<%s
+        ''', (cutoff),
+        )
+        result = cursor.fetchone()['toner_used']
+        # The none check is here as a kind of failsafe for if the request is run on a printer
+        # that is no longer in use, although this should no longer happen in the first place
+        return [None if result is None else float(result)]
```

Comment:
i'm not really a fan of this kind of error handling. it seems like a real error to run this function on a no-longer-used printer, right? if so, i'd expect the function to error when i call it.

generally i'd prefer to fail loud so that we're alerted that something is broken and can fix things. an easy way to do that without having a yucky error message (like `TypeError: float() argument must be a string or a number, not 'NoneType'`) is to just add an assertion before it:

```python
assert result is not None, 'No data exists for printer: {}'.format(printer)
return float(result)
```

Reply:
where would this error be handled?

Reply:
you don't need to handle it if we never expect it to happen. "programming bugs" like these can't really be handled in a sane way except by degrading the experience (e.g. 500ing the page or not showing the affected component) and notifying staff to fix it.

if it *were* to start happening, that would mean there's a bug in ocfweb and we'd automatically start getting emailed about it and know to fix it.